File: complex/src/main/java/net/fwitz/math/numbers/IntSet.java
package net.fwitz.math.numbers;

import java.util.BitSet;
import java.util.NoSuchElementException;
import java.util.PrimitiveIterator;
import java.util.stream.IntStream;

public class IntSet {
    public static IntSet ints(BitSet values) {
        return new IntSet((BitSet) values.clone());
    }

    public static IntSet ints(IntSet values) {
        return new IntSet(values.bits, values.size);
    }

    private final BitSet bits;
    private final int size;

    private int hash;

    private IntSet(BitSet bits) {
        this(bits, bits.length());
    }

    private IntSet(BitSet bits, int size) {
        this.bits = bits;
        this.size = bits.size();
    }

    public PrimitiveIterator.OfInt iterator() {
        return new Itr();
    }

    public int cardinality() {
        return bits.cardinality();
    }

    public int size() {
        return size;
    }

    public IntStream stream() {
        return bits.stream();
    }

    public IntSet truncate(int newSize) {
        if (newSize < 0) {
            throw new IllegalArgumentException("newSize < 0: " + newSize);
        }
        if (newSize >= size) {
            if (newSize > size) {
                throw new IllegalArgumentException("newSize > size: " + newSize + " > " + size);
            }
            return this;
        }
        return new IntSet(bits, newSize);
    }

    @Override
    public int hashCode() {
        int h = hash;
        if (h == 0) {
            h = bits.hashCode() * 31 + size;
            hash = h;
        }
        return h;
    }

    @Override
    public boolean equals(Object o) {
        return o == this || (o instanceof IntSet && equalTo((IntSet) o));
    }

    private boolean equalTo(IntSet other) {
        return size == other.size && hashCode() == other.hashCode() && bits.equals(other.bits);
    }

    @Override
    public String toString() {
        return bits.toString();
    }
    

    private class Itr implements PrimitiveIterator.OfInt {
        private int next = bits.nextSetBit(0);

        @Override
        public int nextInt() {
            final int bit = next;
            if (bit == -1) {
                throw new NoSuchElementException();
            }
            next = bits.nextSetBit(bit + 1);
            return bit;
        }

        @Override
        public boolean hasNext() {
            return next != -1;
        }
    }
}

File: numth/src/main/kotlin/net/fwitz/math/numth/numbers/IntSet.kt
package net.fwitz.math.numth.numbers

import java.util.*

class IntSet private constructor(
    private val bits: BitSet,
    val size: Int = bits.length()
): Sequence<Int> {
    companion object {
        fun bits(bits: BitSet, size: Int = bits.length()): IntSet {
            val copy = bits.clone() as BitSet
            val firstBitToTrash = copy.nextSetBit(size)
            if (firstBitToTrash != -1) copy.clear(firstBitToTrash, copy.length())
            return IntSet(copy, size)
        }
    }

    override operator fun iterator(): PrimitiveIterator.OfInt = Itr()

    val cardinality get() = bits.cardinality()
    val stream get() = bits.stream()
    val seq get() = iterator().asSequence()

    fun orInto(bits: BitSet) = bits.or(this.bits)
    fun andInto(bits: BitSet) = bits.and(this.bits)

    fun truncate(newSize: Int): IntSet {
        require(newSize >= 0) { "newSize < 0: $newSize" }
        if (newSize >= size) {
            require(newSize == size) { "newSize > size: $newSize > $size" }
            return this
        }

        val firstBitOutOfRange = bits.nextSetBit(newSize)
        if (firstBitOutOfRange == -1) return this

        val copy = bits.clone() as BitSet
        copy.clear(firstBitOutOfRange, size)
        return IntSet(copy, newSize)
    }

    private var hash = 0
    override fun hashCode(): Int {
        var h = hash
        if (h == 0) {
            h = bits.hashCode() * 31 + size
            hash = h
        }
        return h
    }

    override fun equals(other: Any?): Boolean {
        return other === this || other is IntSet && equalTo(other)
    }

    private fun equalTo(other: IntSet): Boolean {
        return size == other.size && hashCode() == other.hashCode() && bits == other.bits
    }

    override fun toString() = bits.toString()

    private inner class Itr : PrimitiveIterator.OfInt {
        private var next = bits.nextSetBit(0)

        override fun nextInt(): Int {
            val bit = next
            if (bit == -1) {
                throw NoSuchElementException()
            }
            next = bits.nextSetBit(bit + 1)
            return bit
        }

        override fun hasNext(): Boolean {
            return next != -1
        }

        override fun remove() = throw UnsupportedOperationException()
    }
}